### DISCUSSION TITLE
New mode for navigation

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#1561] [Ideas]
Among the levels of a node, the first level of the node is very important to know in which branch of the first level the desired node is.

A new mode of navigation that I describe can create the most focus for the user to reach the desired node.

New idea: when we go from one node to another node, the node from which the transition was made is folded and the selected node is unfolded one level. In other words: folding of the previous node + unfolding of the first level of the selected node.

The former node from which we moved is the node that we have come to the conclusion that we have nothing to do with it, and the node we are looking for does not exist in its sub-branches, so when we leave it, it is folded and prevents the mental map from being crowded. 
  The selected node is opened one level to identify from which branch to continue the navigation to the desired node.

I hope I have been able to express the idea clearly.
I think this mode of navigation will be useful for many users.

**Comment by aaa1386** [comment]
The idea can be implemented in two ways:

1- Adding four navigation modes for which four shortcuts can be assigned, such as control + four directions (up, down, left, right).

2- Using shift button + left mouse click
Introduction: If the shift or control key is pressed, different nodes can be selected from different places of the mental map by left-clicking the mouse, as shown in the figure below, where the green nodes are selected by the mentioned method.
![image](https://github.com/freeplane/freeplane/assets/38586259/959fa0e3-842b-4077-9c7d-b2f5c258e3ca)

You don't need two methods to select multiple nodes, control + left mouse click is enough
In addition, the function of shift + left click is not complete for selecting multiple nodes because repeating shift + left click does not remove the node from the selection mode.


Therefore, the shift button + left mouse click can be assigned to the mentioned idea.

  If this idea is added to the program, it will not harm the program, but it can be useful for many users.✅✅✅


If the two mentioned methods for this idea are added to the program, the performance will be very complete. If only one of the methods is included, the performance may be reduced for those who are used to using the mouse or shortcut keys.

**Reply by dpolivaev** [reply]
I find the idea interesting and want to give it a try in the coming weeks when I find time. I am going to add a switch "auto-fold" somewhere in the "navigate" menu. When enabled, it should fold / unfold nodes whenever selection changes no matter how.

**Comment by EdoFro** [comment]
> New idea: when we go from one node to another node, the node from which the transition was made is folded and the selected node is unfolded one level. In other words: folding of the previous node + unfolding of the first level of the selected node.

I think it is a good idea. I modified a script of mine to simulate the way it could work, so everybody can test it and have an opinion:

```groovy
import org.freeplane.features.map.INodeSelectionListener;
import org.freeplane.features.map.NodeModel;



def mapController = node.mindMap.modeController.mapController

def myListeners = mapController.nodeSelectionListeners.findAll{it.class.simpleName == 'myNodeSelectionListener'}
println myListeners

//toggles the listener on and off
if(myListeners){
    myListeners.each{mL ->
        mapController.removeNodeSelectionListener(mL)
        mL = null // I don't know if this line is needed.
    }
    ui.informationMessage('listener deactivated')
} else {
    def ptr = node.pathToRoot
    ptr.each{n ->
        (n.children - ptr).each{
            it.setFolded(true)
        }
    }
    mapController.addNodeSelectionListener(new myNodeSelectionListener())
    ui.informationMessage( 'listener activated')
}


class myNodeSelectionListener implements INodeSelectionListener{
    private NodeModel previous;
    
    @Override
    public void onSelect(NodeModel node) {
        node.setFolded(false)
        node.children.each{ n ->
            n.setFolded(true)
        }
        if(!node.pathToRoot.contains(previous))
            previous.setFolded(true)
    }
    
    @Override
    public void onDeselect(NodeModel node) {
        previous = node;
    } 
}
```

In my opinion,
- it should be a view mode (like "Ouline view")
- the folding and unfolding of the nodes in this view shoud only happen in view model level: it shouldn't change the folded state of the nodes itself.
So, when you return to the normal view, the previous folding status should reappear, but the selected node should stay visible (pathToRoot unfolded)

bye,

edo

**Reply by EdoFro** [reply]
EDIT: This script works only with Freeplane version v1.11.9

Here is another version to test.

It works for all selected nodes:

```groovy
import org.freeplane.features.map.INodeSelectionListener;
import org.freeplane.features.map.IMapSelection;
import org.freeplane.features.map.NodeModel;


def mapController = node.mindMap.modeController.mapController
def myListeners = mapController.nodeSelectionListeners.findAll{it.class.simpleName == 'myNodeSelectionListener'}

//toggles the listener on and off
if(myListeners){
    myListeners.each{mL ->
        mapController.removeNodeSelectionListener(mL)
    }
    ui.informationMessage('listener deactivated')
} else {
    def ptr = node.pathToRoot
    ptr.each{n ->
        (n.children - ptr).each{
            it.setFolded(true)
        }
    }
    mapController.addNodeSelectionListener(new myNodeSelectionListener())
    ui.informationMessage( 'listener activated')
}



class myNodeSelectionListener implements INodeSelectionListener{
    private previousSelection = [];
    
    @Override
    public void onSelectionSetChange(IMapSelection mapSelection){
        def selected = [] + mapSelection.selection
        def noMoreSelected = previousSelection - selected
        def newSelected = selected - previousSelection
        
        noMoreSelected.each{m->
            m.pathToRoot.reverse().each{n->
                if(!selected.any{s ->s.isDescendantOf(n)})
                    n.setFolded(true)
            }
        }
        newSelected.each{n->
            n.setFolded(false)
        }
        previousSelection = selected
    }
}
```

**Reply by aaa1386** [reply]
Thank you all, the second script is not working for me.

It is good to implement the idea as a view mode (like "Ouline view").
But besides that, the second method that I proposed to implement the idea(https://github.com/freeplane/freeplane/discussions/1561#discussioncomment-7737113 ) should also be included in the software, 
because many times we don't want to leave the normal view mode, but for example, we need to use the proposed idea once or twice, we can do it by pressing the button. Shift and left mouse click to implement the idea without leaving the normal view mode.


Pay attention to the figure below. I need the first level of the selected node to be open only, not the next levels. The openness of the next levels has made me not have a clear mind about the first level. The above idea can help.
Such things happen sometimes and I don't want to go out of normal view either
![image](https://github.com/freeplane/freeplane/assets/38586259/d3a52e7b-d650-4d05-ac7b-0c6869ddb507)



NODE FOLDED
![image](https://github.com/freeplane/freeplane/assets/38586259/832ae083-d4a2-4e27-9272-15953d4a384a)

**Comment by EdoFro** [comment]
Here is a little video how it looks:


https://github.com/freeplane/freeplane/assets/35700575/e1417e3e-5b86-40a9-aac0-bf95e37caff1

**Reply by EdoFro** [reply]
I made a video for the second script, so you can see how it works.

It's very similar than the first one, but you can also select multiple nodes and they will unfold one level too.


https://github.com/freeplane/freeplane/assets/35700575/005232b3-42dd-4130-a589-fe3b0a2a911b

**Comment by Illegit** [comment]
Hi EdoFro, I tried your 2 scripts, but the 2nd one is generating a null error during execution. instead the 1st is working.
I think that the traditional way provides easier control for editing and this new better for viewing, but in any case it could be nice to have a switch to choose from time to time what to use in both modes. Interesting tool!

**Reply by EdoFro** [reply]
I tried the 2nd again and I have no errors. 🤷🏼

**Reply by EdoFro** [reply]
can you  please include the log.0 file?
I hope there is more information about the error and the line of the script where it happens

**Reply by i-plasm** [reply]
I could be wrong, but the error may have to do with the `  INodeSelectionListener.onSelectionSetChange ` method - it apparently was only added with a commit two weeks ago, thus unavailable to most Freeplane installations -.  @EdoFro  this is a useful navigation method (thank you!), and I agree it would be even better to have the option to have it as a map view, so that the map folding is not changed.

**Reply by EdoFro** [reply]
OH!!! thanks, I didn't knew that.
I'm using a very updated Freeplane version, because I was developing and build it myself, but I didn't realize that I was using a fresh new method for this script.

I edited the second script, so it works better when you can finally test it (if you want).

I also added a second video that shows how it works in my computer.

**Reply by Illegit** [reply]
Here is my log, I confirm again the null error with this script.
[log (2).zip](https://github.com/freeplane/freeplane/files/13549575/log.2.zip)

**Reply by Illegit** [reply]
Hi EdoFro with your 1st script I immediately activated this MoM menu: 
![image](https://github.com/freeplane/freeplane/assets/62674157/241d843f-b74a-4566-b060-1110adead89a)

**Reply by i-plasm** [reply]
Hi @EdoFro . On sourceforge the latest Freeplane version was published Nov 12, whereas the commit I mentioned is dated Nov 16 (https://github.com/freeplane/freeplane/commit/dfa114548d6e2496a85e2d07ae5ac14de51e2051) ... so the second script cannot be tested with user packages yet. If I set a new workspace to the latest Freeplane code I'll be able to try it - currently my IDE has Freeplane v1.9.12 code (for backwards compat), go figure!

**Comment by aaa1386** [comment]
In my opinion, if the following 3 items are added to the program, it can satisfy the opinions of all users and the software will be more complete in terms of navigation.
1- The proposed idea should be added to the map view for those who do not want to have any changes in the opened or unopened nodes according to their previous settings.

2- The idea presented in the normal view and without moving to the map view mode: pressing the shift button and then selecting the desired node by clicking the mouse (shift + left click) - for those who sometimes navigate in this mode while working with the map. They need it. These people like the map to be folded according to the idea. Of course, the first level should only be folded and the subsequent levels should not be manipulated according to the user's settings. For this purpose, when the previous node is folded, the software should not make any changes to the settings of the second and third nodes, etc., due to whether the nodes are open or closed, but only the first level should be closed. Therefore, the least manipulation happens in the user's previous settings.
Pay attention to the figure below, the first part of the idea must be exactly in accordance with the function "Navigit>Fold>(un)Fold" when it performs the fold, that is, the settings of the user in the open or closed nodes of the branches, the minister of branches remembers. The second part of the idea that opens the selected node by one level should also be in accordance with the "Navigate>Fold>(un)Fold" function, except that it keeps the first level open but closes the next levels. The previous settings of the game should store the closure of other levels by the user.
In general: the open or closed settings software of branches and sub-branches should be remembered when implementing the idea, similar to the function of the software in "Navigit>Fold>(un)Fold"
"Navigate>Fold All" function should not happen when the idea is executed by the software.
![image](https://github.com/freeplane/freeplane/assets/38586259/93b16d70-9b06-4b3f-9340-63655438036a)


3- Adding a new navigation mode to fold the node except for the first layer, just like the navigation>fold>(un)fold function, the only difference is that it keeps the first layer open. Of course, this idea is new and has nothing to do with the idea presented in this thread.

**Reply by aaa1386** [reply]
In my opinion, a fourth action needs to be done:
The implementation of the idea was explained for when the user moves from one node to another using the mouse. But when moving between the nodes with the arrow keys (up, down, right, left), four navigation modes should be created:
1-Select Up Node + Auto Fold (this navigation mode both selects the top node and executes the idea)
2- Select the bottom node + auto fold (this navigation mode both selects the bottom node and executes the idea)
3- Select the right node + auto fold (this navigation mode both selects the right node and executes the idea)
4-Select the left node + auto fold (this navigation mode both selects the left node and executes the idea)
Users can define four shortcuts for these four modes, such as:
  shift + ←
Shift + →
shift + up
shift + down

Auto-Fold can also be defined in two modes in order to make the idea more complete. It needs a check box. The user can choose: 1- Folding all the nodes from the map + Unfolding the first level of the selected node.
2- Folding the previous node + unfolding the first level of the selected node.

It was mentioned several times that folding the node when implementing the idea by the software should not mess up the settings of the branch nodes that the user has folded or unfolded, it only changes the first level.
Because the least changes in folded nodes or unfolded nodes occur in branches and sub-branches. User settings are not disturbed as much as possible.

**Reply by aaa1386** [reply]
If it is possible to implement these 4 actions in the software, it seems that good things will happen for the software.

**Reply by aaa1386** [reply]
If all the items of this idea are implemented in accordance with all the details I wrote, a big upgrade will happen in the software.Others may not believe as much.

**Comment by adamfreeplaneuser01** [comment]
Unless I'm misunderstanding, I brought up/asked for a similar feature here: #1464. However, you are articulating the idea in a way that actually makes sense lol, whereas I wasn't sure how to actually present the request

**Comment by dpolivaev** [comment]
Implemented
Preview 1.11.9_03
Navigate->Auto-expand selected nodes

**Reply by dpolivaev** [reply]
As per-view option

**Reply by aaa1386** [reply]
1- Activate the "Auto Expand" option.
2- Then go to root.
3-Expand root and all its children.(unfold all)

After doing these steps, "Auto Expand" does not work well.
Even for fully expanded nodes and branches, the "Auto Expand" option does not work well.

[bug.zip](https://github.com/freeplane/freeplane/files/13773365/bug.zip)

**Reply by dpolivaev** [reply]
Dear @ajaberi1360,

Thank you for your message and for bringing this issue to our attention. I understand your concern regarding the "Auto Expand" feature. To clarify, the feature is designed to fold only those nodes that it unfolded earlier, while leaving other nodes in their current state, either folded or unfolded.

Additionally, I'd like to gently point out something regarding the tone of bug reports. Starting a complaint with phrases like "Thanks to the software developer" can sometimes be interpreted differently than intended. While I understand this may not have been your intention, such phrases can occasionally come across as sarcastic. This might inadvertently impact the constructive nature of our communication. Please know that I am deeply committed to resolving any issues and improving the user experience, and I believe that clear, straightforward communication is key to achieving this effectively.

Thank you again for your feedback. If you have any more details or specific scenarios where the "Auto Expand" feature does not perform as expected, please do not hesitate to share them. Your input is invaluable in helping us enhance our software.

Best regards,
Dimitry

**Comment by aaa1386** [comment]


**Reply by dpolivaev** [reply]
I do not understand what is different in the "second" and the "third" mode.
The mode I implemented is flexible and fits all cases because you can always fold all nodes of the map, unfold them, fold only next-level children, and unfold them. Combining it with the auto-expand you can achieve any behavior I can imagine, therefore in my eyes, my solution satisfies all needs.

**Reply by dpolivaev** [reply]
If you select the root node and fold all nodes (Navigate->Fold->Fold all) before you start navigating with auto-expand, it should give you what you describe. Doesn't it?

**Reply by aaa1386** [reply]
I worked more with the new changes, my opinion has changed compared to some previous comments. I deleted them.
With the special flexibility that "Auto Expand" has, it fulfills the need.
✅✅

**Reply by aaa1386** [reply]
thanks.

**Comment by EdoFro** [comment]
> Implemented 
> Preview 1.11.9_03 
> Navigate->Auto-expand selected nodes

Dear @dpolivaev,
I have just tested this new functionality. I really like the way it is set up. 👏🏼 🍾 
Thanks again!
